File: backends/python-fastapi/app/domain/invitation/repository.py
"""
招待ドメイン リポジトリインターフェースモジュール。

Author: Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
"""
from abc import ABC, abstractmethod
from typing import Optional

from app.domain.invitation.value_objects import InvitationVo


class InvitationRepository(ABC):
    """招待の永続化インターフェース。"""

    @abstractmethod
    def get_current_by_role(self, role: int) -> Optional[InvitationVo]:
        """ロールで絞り込んだ最新の招待情報の Vo を返します。

        Args:
            role: ロール（1=管理者, 2=メンバー）

        Returns:
            InvitationVo、または None
        """
        ...

    @abstractmethod
    def issue(self, role: int) -> InvitationVo:
        """新しい招待トークンを生成して保存し、Vo を返します。

        Args:
            role: ロール（1=管理者, 2=メンバー）

        Returns:
            InvitationVo インスタンス
        """
        ...

    @abstractmethod
    def find_by_token(self, token: str) -> Optional[InvitationVo]:
        """トークンで招待情報の Vo を返します。存在しない場合は None を返します。

        Args:
            token: 招待トークン

        Returns:
            InvitationVo、または None
        """
        ...

File: backends/ruby-rails/app/domain/invitation/repository.rb
# frozen_string_literal: true
#
# 招待リポジトリインターフェースを定義するドメインオブジェクトモジュール。
#
# @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

module Domain
  module Invitation
    # 招待リポジトリのインターフェースです。
    module Repository
      # @param role [Integer] ロール（1=管理者, 2=メンバー）
      # @return [Domain::Invitation::Entity, nil] 現在有効な招待エンティティ
      def get_current_by_role(role) = raise NotImplementedError

      # @param role [Integer] ロール（1=管理者, 2=メンバー）
      # @return [Domain::Invitation::Entity] 発行した招待エンティティ
      def issue(role)               = raise NotImplementedError

      # @param token [String] 招待トークン
      # @return [Domain::Invitation::Entity, nil] 招待エンティティ
      def find_by_token(token)      = raise NotImplementedError
    end
  end
end
